### ISSUE TITLE
ansible_ssh_retries parameter does not work for delegated connection (ansible ver 2.9.23)

### ISSUE THREAD

**Issue by canbahadir** [ansible/ansible#77149]
### Summary

When I try to configure ssh retry parameter through vars in a delegated task, it does not behave correctly. Default retry parameter is 3 which works as expected when not configured, but when I add ansible_ssh_retries with any number value, no chance happens.

### Issue Type

Documention Report

### Component Name

ssh

### Ansible Version

```console
$ ansible --version
ansible 2.9.23
  config file = /examples/ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /opt/venv/lib/python3.6/site-packages/ansible
  executable location = /opt/venv/bin/ansible
  python version = 3.6.8 (default, Sep  9 2021, 07:49:02) [GCC 8.5.0 20210514 (Red Hat 8.5.0-3)]
```

### Configuration

```console
$ ansible-config dump --only-changed
ANSIBLE_SSH_RETRIES(/examples/ansible.cfg) = 10
```

### OS / Environment

$ cat /etc/redhat-release 
Red Hat Enterprise Linux release 8.4 (Ootpa)

### Steps to Reproduce

ansible.cfg:

```
[ssh_connection]
retries = 10
```

retry_test.yml:

```
- name: "Create a file"
  hosts: localhost
  gather_facts: no
  any_errors_fatal: true
  tasks:
    - name: "Create a file if not present"
      delegate_to: <host_ip>
      vars:
        ansible_user:  <user>
        ansible_ssh_pass: <pw>
        ansible_become_method: su
        ansible_become: yes
        ansible_become_password: <pw>
        ansible_ssh_retries: 10
        ansible_ssh_extra_args: '-o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null'
      file:
        path: "/home/<user>/randomfile.txt"
        state: touch

```

Run following cmd while delegated host ssh connection limit defined by /etc/security/limits.conf is exceeded:

`ansible-playbook -i hosts.yml retry_test.yml -vvv`

### Expected Results

When ansible_ssh_retries added to vars with a number value X, ssh should try up to X attempts to connect for delegated_hosts.

### Actual Results

```console
[root@d94f53b2a6bb:]$ ansible-playbook -i hosts.yml retry_test.yml -vvv
ansible-playbook 2.9.23
  config file = /examples/ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /opt/venv/lib/python3.6/site-packages/ansible
  executable location = /opt/venv/bin/ansible-playbook
  python version = 3.6.8 (default, Sep  9 2021, 07:49:02) [GCC 8.5.0 20210514 (Red Hat 8.5.0-3)]
Using /examples/ansible.cfg as config file
host_list declined parsing /examples/hosts.yml as it did not pass its verify_file() method
script declined parsing /examples/hosts.yml as it did not pass its verify_file() method
Parsed /examples/hosts.yml inventory source with yaml plugin
Skipping callback 'actionable', as we already have a stdout callback.
Skipping callback 'counter_enabled', as we already have a stdout callback.
Skipping callback 'debug', as we already have a stdout callback.
Skipping callback 'dense', as we already have a stdout callback.
Skipping callback 'dense', as we already have a stdout callback.
Skipping callback 'full_skip', as we already have a stdout callback.
Skipping callback 'json', as we already have a stdout callback.
Skipping callback 'minimal', as we already have a stdout callback.
Skipping callback 'null', as we already have a stdout callback.
Skipping callback 'oneline', as we already have a stdout callback.
Skipping callback 'selective', as we already have a stdout callback.
Skipping callback 'skippy', as we already have a stdout callback.
Skipping callback 'stderr', as we already have a stdout callback.
Skipping callback 'unixy', as we already have a stdout callback.
Skipping callback 'yaml', as we already have a stdout callback.

PLAYBOOK: retry_test.yml ****************************************************************************************************************************************************************
1 plays in retry_test.yml

PLAY [Create a file] ********************************************************************************************************************************************************************
META: ran handlers

TASK [Create a file if not present] *****************************************************************************************************************************************************
task path: /examples/retry_test.yml:6
<<host_ip>> ESTABLISH SSH CONNECTION FOR USER: <host_user>
<<host_ip>> SSH: EXEC sshpass -d10 ssh -C -o ControlMaster=auto -o ControlPersist=60s -o 'User="<host_user>"' -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o ControlPath=/root/.ansible/cp/fbfe67135b <host_ip> '/bin/sh -c '"'"'echo ~<host_user> && sleep 0'"'"''
<<host_ip>> (254, b"\nAuthorized users only. All activity may be monitored and reported.\n\nToo many logins for '<host_user>'.\n", b"Warning: Permanently added '<host_ip>' (ECDSA) to the list of known hosts.\r\n\nAuthorized users only. All activity may be monitored and reported.\n\n")
<<host_ip>> Failed to connect to the host via ssh: Warning: Permanently added '<host_ip>' (ECDSA) to the list of known hosts.

Authorized users only. All activity may be monitored and reported.
<<host_ip>> ESTABLISH SSH CONNECTION FOR USER: <host_user>
<<host_ip>> SSH: EXEC sshpass -d10 ssh -C -o ControlMaster=auto -o ControlPersist=60s -o 'User="<host_user>"' -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o ControlPath=/root/.ansible/cp/fbfe67135b <host_ip> '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo Too many logins for '"'"'"'"'"'"'"'"'<host_user>'"'"'"'"'"'"'"'"'./.ansible/tmp `"&& mkdir "` echo Too many logins for '"'"'"'"'"'"'"'"'<host_user>'"'"'"'"'"'"'"'"'./.ansible/tmp/ansible-tmp-1646032840.9470909-227-194794013968900 `" && echo ansible-tmp-1646032840.9470909-227-194794013968900="` echo Too many logins for '"'"'"'"'"'"'"'"'<host_user>'"'"'"'"'"'"'"'"'./.ansible/tmp/ansible-tmp-1646032840.9470909-227-194794013968900 `" ) && sleep 0'"'"''
<<host_ip>> (254, b'', b'')
<<host_ip>> Failed to connect to the host via ssh: 
fatal: [localhost]: UNREACHABLE! => {
    "changed": false,
    "msg": "Failed to create temporary directory.In some cases, you may have been able to authenticate and did not have permissions on the target directory. Consider changing the remote tmp path in ansible.cfg to a path rooted in \"/tmp\", for more error information use -vvv. Failed command was: ( umask 77 && mkdir -p \"` echo Too many logins for '<host_user>'./.ansible/tmp `\"&& mkdir \"` echo Too many logins for '<host_user>'./.ansible/tmp/ansible-tmp-1646032840.9470909-227-194794013968900 `\" && echo ansible-tmp-1646032840.9470909-227-194794013968900=\"` echo Too many logins for '<host_user>'./.ansible/tmp/ansible-tmp-1646032840.9470909-227-194794013968900 `\" ), exited with result 254",
    "unreachable": true
}

NO MORE HOSTS LEFT **********************************************************************************************************************************************************************

PLAY RECAP ******************************************************************************************************************************************************************************
localhost                  : ok=0    changed=0    unreachable=1    failed=0    skipped=0    rescued=0    ignored=0
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by ansibot** [comment]
Files identified in the description:
* [`lib/ansible/plugins/connection/ssh.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/connection/ssh.py)

If these files are incorrect, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/devel/ISSUE_HELP.md)

**Comment by OUARO** [comment]
Bonjour 
J'ai une version 2.7.10 de ansible mais je voudrais fixer sur la nouvelle version de ansible. 
Quelle commande il me faut?
voici le message d'erreur. Je vous remercie.
**![erreur de version de ansible](https://user-images.githubusercontent.com/72983718/155979628-1c2173ba-a3e0-4c2f-805c-eb7aeb638f66.PNG)**

**Comment by mkrizek** [comment]
This does not appear to be related to delegation but to the ssh `retries` option in general.

> Default retry parameter is 3 which works as expected when not configured, ...

This does not appear to be correct, the default is 0 as [documented](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/ssh_connection.html#parameter-reconnection_retries).

The issue is in `_handle_error` function in the `ssh` connection plugin, specifically in https://github.com/ansible/ansible/blob/68601601ce487a543c92ae3e179acc478666ba6a/lib/ansible/plugins/connection/ssh.py#L449 not raising an exception does not imply retrying the connection as we stop retry in https://github.com/ansible/ansible/blob/68601601ce487a543c92ae3e179acc478666ba6a/lib/ansible/plugins/connection/ssh.py#L507

**Comment by mkrizek** [comment]
The connection is only retried for "SSH errors" (the return code of 255). This is working as intended.

Changing this to a Documention Report. We should mention this in https://docs.ansible.com/ansible/latest/collections/ansible/builtin/ssh_connection.html#parameter-reconnection_retries.

**Comment by canbahadir** [comment]
First off, yes retry default is 0. I mistaken different ssh cmds as retries on logs. So I don`t get retries for create file task beacuse denying due limit of ssh connection return code is 254 as you already mention.


So I tried it with another task that I first found the case, wait_for_connection.

I have following ymls this time: 

retry_test.yml

```
- name: "Wait for system to become reachable"
  hosts: remote_host
  gather_facts: no
  any_errors_fatal: true
  tasks:
    - name: Wait for system to become reachable
      wait_for_connection:
        timeout: 40
```

retry_test_delegated.yml

```
- name: "Wait for system to become reachable"
  hosts: localhost
  gather_facts: no
  any_errors_fatal: true
  tasks:
    - name: "Wait for system to become reachable"
      delegate_to: <host_ip>
      vars:
        ansible_user:  <user>
        ansible_ssh_pass: <pw>
        ansible_become_method: su
        ansible_become: yes
        ansible_become_password: <pw>
        ansible_ssh_retries: 10
        ansible_ssh_extra_args: '-o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null'
      wait_for_connection:
        timeout: 50
```

hosts.yml

```
all:
  hosts:
    remote_host:
      ansible_host: <host_ip>
      ansible_user: <user>
      ansible_ssh_pass: <pw>
      ansible_become_method: su
      ansible_become: yes
      ansible_become_password: <pw>
      ansible_ssh_extra_args: '-o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null'
```



some tests I applied:

```
wait for connection timeout:30 | host with retry:10                  | retry appeared at logs  5 times          | took 48 sec
wait for connection timeout:40 | host with retry:10                  | retry appeared at logs  6 times          | took 97 sec
wait for connection timeout:40 | delegated host with retry:10        | retry did not appeared at logs           | took 40 sec
wait for connection timeout:30 | delegated host with retry:10        | retry appeared at logs  6 times          | took 87 sec
wait for connection timeout:50 | delegated host with retry:10        | retry appeared at logs  2 times          | took 51 sec
wait for connection timeout:50 | delegated host with retry:10        | retry appeared at logs  5 times          | took 50 sec
```



For wait_for_connection ssh sometimes gets return code of 255 it seems.

I also would like to ask this: Can [wait_for_connection + desired_task] be a solution for retrying ssh connections for such cases like this where return code is not 255 and multiple connections need to be made in parallel?

e.g x has 1 connection limit and a,b an c tries to connect x at the same time. 
lets say a connected x first then b an c waits. a finished task and disconnected, b gets signal with wait for connection and tries to run next task while c also gets connection in between (assuming connection does not persist between wait_for_connection and desired_task )

Could this also fail for this case?


Extensive log of tests:

https://gist.github.com/canbahadir/97f7365b5160581f7fb36d958623f205
